pkg/kubevirt/utils.go

```diff
@@ -138,36 +143,90 @@ func buildVirtualMachineInstanceTemplate(ctx *context.MachineContext) *kubevirtv
 
 	template.Spec = *ctx.KubevirtMachine.Spec.VirtualMachineTemplate.Spec.Template.Spec.DeepCopy()
 
-	cloudInitVolumeName := "cloudinitvolume"
-	cloudInitVolume := kubevirtv1.Volume{
-		Name: cloudInitVolumeName,
-		VolumeSource: kubevirtv1.VolumeSource{
-			CloudInitConfigDrive: &kubevirtv1.CloudInitConfigDriveSource{
-				UserDataSecretRef: &corev1.LocalObjectReference{
-					Name: *ctx.Machine.Spec.Bootstrap.DataSecretName + "-userdata",
+	cloudInitType, index := detectCloudInitVolumeType(template)
+	cloudInitVolume := cloudinitVolume(template, ctx, cloudInitType)
+	cloudInitDisk := cloudinitDisk(template, cloudInitType)
+	switch cloudInitType {
+	case "CloudInitNoCloud":
+		err := mergo.Merge(&template.Spec.Volumes[index], cloudInitVolume)
+		if err != nil {
+			panic(err)
```

Comment:
Panic will kill the whole process. Are uou sure this is the behaviour we want here?

Reply:
_buildVirtualMachineInstanceTemplate_ doesn't return an error.
Which behaviour would you suggest here?

Reply:
I'm unsure how the behaviour would be if it were unhandled, like

```go
		_ = mergo.Merge(&template.Spec.Volumes[index], cloudInitVolume)
```

but perhaps that's an option?

Reply:
After reading the code of `mergo.Merge`, I think it's too risky to ignore the error. I'm afraid there is no other option but change the signature of the `buildVirtualMachineInstanceTemplate` function, to add an error return value, and to do the same for the calling function, `newVirtualMachineFromKubevirtMachine`, then check this error in the `Machine::Create` methods (that does return error). These functions are also called from unit tests, so there is no choice but fixing these test as well.

Reply:
Fair evaluation! 
Updated in bb66a2d